clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/ReconnectingServerStoreProxy.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.clustered.client.internal.store;
+
+import org.ehcache.clustered.common.internal.store.Chain;
+import org.terracotta.exception.ConnectionClosedException;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ReconnectingServerStoreProxy implements ServerStoreProxy {
+
+  private final ServerStoreProxy swap;
+  private final ServerStoreProxy delegate;
+  private final AtomicReference<ServerStoreProxy> delegateRef;
+  private final Runnable onReconnect;
+
+  public ReconnectingServerStoreProxy(ServerStoreProxy serverStoreProxy, Runnable onReconnect) {
+    this.delegate = serverStoreProxy;
+    this.delegateRef = new AtomicReference<>(serverStoreProxy);
+    this.onReconnect = onReconnect;
+    this.swap = new SwappableProxy(serverStoreProxy.getCacheId());
```

Comment:
Bunch of things seem weird here.

1. Why do we need a create this thing now?  Why not just do it on demand?
2. Why is it called SwappableProxy, isn't it really `ReconnectInProgressProxy`?
3. Lots of code duplication here.

Wondering if we can solve all this with a method something like this:
```java
  private <T> T onStoreProxy(Function<ServerStoreProxy, T> function) {
    ServerStoreProxy storeProxy = proxy();
    try {
      return function.apply(storeProxy);
    } catch (ServerStoreProxyException sspe) {
      if (sspe.getCause() instanceof ConnectionClosedException) {
        if (delegateRef.compareAndSet(storeProxy, new ReconnectInProgress(storeProxy.getCacheId()))) {
          onReconnect.run();
        }
        return onStoreProxy(function);
      } else {
        throw sspe;
      }
    }
  }
```

Reply:
This is cool